addons/web/static/src/js/views/file_upload_mixin.js

```diff
@@ -0,0 +1,217 @@
+        let title = _.str.sprintf(_t('%s Files'), files.length);
+        let type;
+        if (files.length === 1) {
+            title = files[0].name;
+            type = files[0].type;
+        }
+
+        await this._uploadFiles('/web/binary/upload_attachment', files, {
+            onUpload: async xhr => {
+                this._makeFileUploadProgressBar(fileUploadId, {xhr, title, type});
+                await this._attachFileUploadProgressBars();
+            },
+            onload: () => {
+                this._removeUpload(fileUploadId);
+            },
+            onerror: xhr => {
+                this.do_notify(xhr.status, _.str.sprintf(_t('message: %s'), xhr.reponseText), true);
+                this._removeUpload(fileUploadId);
+            },
+            formData: {
+                callback: fileUploadId,
+            }
+        });
+    },
+    /**
+     * @private
+     * @param {integer} fileUploadId
+     */
+    async _removeUpload(fileUploadId) {
+        const upload = this._uploads[fileUploadId];
+        upload.progressCard && upload.progressCard.destroy();
+        upload.progressBar.destroy();
+        delete this._uploads[fileUploadId];
+        await this.reload();
+    },
+    /**
+     * @private
+     * @param {String} route
+     * @param {Object[]} files
+     * @param {Object} [params] optional additional data
+     * @param {function} [params.onUpload] called before sending the xhr
+     * @param {function} [params.onload] called on xhr.onload
+     * @param {function} [params.onerror] called on xhr.onerror
+     * @param {Object} [params.formData] values to be set in the upload's FormData (args of the route).
+     */
+    async _uploadFiles(route, files=[], params={}) {
+        if (!files.length) { return; }
+
+        const formData = new FormData();
+        formData.append('csrf_token', csrf_token);
+        for (const key in params.formData) {
+            if (params.formData[key] !== undefined) {
+                formData.append(key, params.formData[key]);
+            }
+        }
+        for (const file of files) {
+            formData.append('ufile', file);
+        }
+
+        await new Promise(resolve => {
+            const xhr = this._createXhr();
+            xhr.open('POST', route);
+            const progressPromise = params.onUpload && params.onUpload(xhr);
+            xhr.onload = async () => {
+                await progressPromise;
+                resolve();
+                params.onload && params.onload(xhr);
+            };
+            xhr.onerror = async () => {
+                await progressPromise;
+                resolve();
+                params.onerror && params.onerror(xhr);
+            };
+            xhr.send(formData);
+        });
+    },
+
+    //--------------------------------------------------------------------------
+    // Handlers
+    //--------------------------------------------------------------------------
+
+    /**
+     *
```

Comment:
start with tags if no comment (that's okay for odoo event handlers that are straight-forward like this one)